src/network/token_provider.rs

```diff
@@ -64,14 +68,76 @@ impl TokenProvider for IBMCloudTokenProvider {
         form_data.insert("apikey".to_string(), self.apikey.to_string());
 
         let client = Client::new();
-        Ok(client
+        let new_token = client
             .post("https://iam.cloud.ibm.com/identity/token")
             .header("Accept", "application/json")
             .form(&form_data)
             .send()
             .map_err(NetworkError::ReqwestError)?
             .json::<AccessTokenResponse>()
-            .map_err(NetworkError::ReqwestError)? // FIXME: This is a deserialization error (extract it from Reqwest)
-            .access_token)
+            .map_err(NetworkError::ReqwestError)?; // FIXME: This is a deserialization error (extract it from Reqwest)
+
+        let mut access_token = self.access_token.write()?;
+        access_token.renew(new_token.access_token, new_token.expires_in);
+        Ok(())
+    }
+}
+
+#[derive(Deserialize)]
+struct AccessTokenResponse {
+    access_token: String,
+    expires_in: u64,
+    // We are discarding some fields:
+    // refresh_token: String, // "not_supported"
+    // token_type: String,    // "Bearer"
+    // expiration: u64,       // 1755854106 <- unix time when it expires
+    // scope: String,         // "ibm openid"
+}
+
+impl TokenProvider for IBMCloudTokenProvider {
+    fn get_access_token(&self) -> NetworkResult<String> {
+        if self.access_token.read()?.expired() {
+            self.renew_token()?;
+        }
+
+        Ok(self.access_token.read()?.token.to_string())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_access_token() {
+        let mut access_token = AccessToken::default();
+        assert!(access_token.expired());
+
+        access_token.renew("something".to_string(), 10);
+        assert!(!access_token.expired());
+        assert_eq!(access_token.token, "something".to_string());
+    }
+
+    #[test]
+    fn test_ibm_cloud_token_provider() {
+        let provider = IBMCloudTokenProvider::new("apikey");
+        assert!(provider.expired());
+
+        // If the token is expired, it will try to renew it when requesting it
+        assert!(matches!(
+            provider.get_access_token().unwrap_err(),
+            NetworkError::ReqwestError(_)
+        ));
```

Comment:
This is already quite a good test, but it does not test any deserialization and also not if the actual renewal is performed.

Reply:
The way it is implemented today (everything is hardcoded inside `IBMTokenProvider::get_access_token`) it will require to run an actual server and somehow redirect the URL (`https://iam.cloud.ibm.com/identity/token`), right? It will require some refactor so we can provide this data from outside.

I can do this refactor and use `httpmock`, but in a different PR